### ISSUE TITLE
Hanging fluid during some MinGW (MSYS2) builds

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#635]
Under some conditions users (including myself) have observed "hanging" `fluid` or `fluid-cmd` processes during the build process of FLTK under MSYS2. The details are not really clear.

In my own environment I had this issue for a long time but since I use my Windows 10 + MSYS2 (mingw-w64) build only for testing FLTK I didn't try to track this down. One symptom was that the build process seemed to background itself (the shell prompt appeared immediately and ctrl/c didn't work - obviously because the process was in the background).

The workaround in this constellation was to kill the hanging fluid processes via task manager and to restart the build. After one or more tries it succeeded. In my case the "hang" didn't always happen at the same point of the build, it looked like a random issue. Maybe related to virus checker activities?

Some time later I discovered that I had a PATH issue and used the Windows GUI `CMake` program inside the MSYS2 shell environment. This led to obscure problems. The reason for this problem was an incomplete MSYS installation - which I fixed later.

After the above mentioned fix of my MSYS2 installation I do no longer experience this "hanging fluid" issue in my own build system.

Other users reported that they had similar effects. We need to investigate what can cause these "hangs" and how to install MSYS2 reliably in a way that it doesn't produce these hanging fluid processes.

Further information can be found in the long discussion (after the original problem) in issue #445 ("type issue in fluid/ExternalCodeEditor_WIN32.cxx") starting with [this comment](https://github.com/fltk/fltk/issues/445#issuecomment-1133841647): "now the build process is stuck at Generating fast_slow.cxx, fast_slow.h ...".

**Comment by J0hanss0n** [comment]
I did some tests that afternoon (triggered by this newly opened issue).
I observed the issue with autoconf/configure/make on MSYS2 (MSYS_NT-10.0-19044) in a MINGW32 Shell. So it is not related to CMake in my case.
But:
With current master (69773338b) the issue is gone (nothing updated/changed in my MSYS2 installation).
And:
As soon as I checkout the version then (531ab614e) the issue is reproducible there. I could do a git bisect, but each build takes several minutes under MSYS2 on my "machine"...

**Comment by MatthiasWM** [comment]
The commit removes at least two places where FLUID may try to connect to the display, which it should not. I am not familiar enough with Windows MSYS to know why that would randomly hang.

**Comment by Albrecht-S** [comment]
@J0hanss0n Thank you for testing and bisecting. This is very much appreciated and helpful. I'm just starting my Windows VM for a test to see if I can verify this.

@MatthiasWM Yes, I've seen that too and I don't know either why this would hang but I'm now trying to reproduce the issue.

I'll let you know what I find out shortly...

**Comment by Albrecht-S** [comment]
OK, I can confirm @J0hanss0n's findings: commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de is the first working commit, i.e. one that doesn't "hang" in `fluid-cmd`. Thanks again, that's a great result!

I also tested whether it has to do with the new console version `fluid-cmd` versus the older GUI version `fluid` used to build the library and test programs. This change in commit 228d04d0e0d6e968c59bcab606e609eb1103f917 didn't affect the "hanging" though.

In my own builds the "unintended fix" in commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de and my updates of the MSYS2 build environment must have crossed in time which made me think it might have to do with the build environment. I can now rule this out as well.

@MatthiasWM It **might** have to do something with the contents of the fluid (`.fl`) file as we had seen in another case (which was, IIRC, parsing a color name on X11).

That said, now I'm going to try to track the issue down (even if it's already "fixed") to see **why** the build process hung in `fluid` or later in `fluid-cmd`.

Going to investigate further...

**Comment by Albrecht-S** [comment]
First result: Windows says it's a deadlock:

![image](https://user-images.githubusercontent.com/26235500/211073765-5968e448-618c-4cf4-9582-73181b5ae8e7.png)

I have no idea what these threads are about. Investigating further ...

**Comment by erco77** [comment]
So I'm just watching this from afar, but thought I'd mention: If you can replicate the hang, I would think you could use Visual Studio to attach to the process via the PID, and then tell it to break and look at the calling stack to see what's going on.

If it still hangs with debugging enabled, all the better as you'd be able to see what line in the code is hanging.

Pretty sure you can debug processes even if they weren't built with VS (in this case mingw), or perhaps mingw includes the gnu gdb(1) tool which will also let you attach to an already running process, break it, and inspect the stack with 'where'.

**Comment by MatthiasWM** [comment]
> @MatthiasWM It **might** have to do something with the contents of the fluid (`.fl`) file as we had seen in another case (which was, IIRC, parsing a color name on X11).

Neither a FLUID nor an .fl file was changed in that commit. But you could undo the changes that remove the code to open the display connection, and see if it hangs again?! Best of course is always the debugger.

**Comment by Albrecht-S** [comment]
@erco77 Thanks for chiming in with these important hints, your help is very much appreciated. I'm not used to Windows debugging and I'll try how far I can get with the debugger, attaching to a running process. We'll see...

Meanwhile I made some interesting observations. My test procedure in my CMake build environment is:
```
$ rm -rf test/*.cxx test/*.h bin/test/*.exe; time ninja -j<n>
```

... where \<n\> varies. If I had configured to use `make` it would be `make -j<n>` instead. The `rm` command cleans all generated .cxx/.h files and all executables (might be faster if I deleted only those that use .fl files, but anyway, that's what I'm doing now).

Running this commandline I can see in the task manager that I get about n-1 or n-2 "hanging" fluid processes. I can see the commandline (in process details, after enabling this column), and it doesn't seem to have to do with the particular build target that hangs, it seems more to do with how many parallel jobs are running.

If I run with `-j1` or `-j2` the build **seems** to complete reliably [1], the problem starts with `-j3`.

[1] Correction, after running often enough I could also reproduce a hang with `-j2`.

- [ ] Is it a Windows resource problem?
- [ ] Can it be that different processes cause deadlocks of each other?

I **_like_** Windows! :-(

In German I call it "Windoof" which is [pronounced with a long 'O'](https://dict.leo.org/englisch-deutsch/doof), not like in "school" (see link for pronounciation and translation). That German word combination says a lot ;-)

Dinner break!

**Comment by Albrecht-S** [comment]
> Neither a FLUID nor an .fl file was changed in that commit ...

Thanks, yes, sure, I know. But it could have been two factors that need to come together.

> ... you could undo the changes that remove the code to open the display connection, and see if it hangs again

Yes, I'll try that too (if possible and/or necessary). My first attempt was to see how it can be reproduced reliably and I my results seem to exclude the fl file contents - because the hang is caused by almost all files or none, depending on the parallel execution.

Thanks for all help!

As I wrote before, it's dinner time...

**Comment by Albrecht-S** [comment]
More results from gdb, looking around w/o a real plan, but ...
```
(gdb) bt
# 0  0x00007ffdf382d144 in ntdll!ZwWaitForSingleObject () from C:\Windows\SYSTEM32\ntdll.dll
# 1  0x00007ffdf0ea306e in WaitForSingleObjectEx () from C:\Windows\System32\KernelBase.dll
# 2  0x00007ffdcf5c8efa in gdiplus!GdiplusStartup ()
   from C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll
# 3  0x00007ffdcf5c6bc4 in gdiplus!GdiplusShutdown ()
   from C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll
# 4  0x00007ffdcf5c6a43 in gdiplus!GdiplusShutdown ()
   from C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll
# 5  0x00007ff7d992843b in Fl_GDIplus_Graphics_Driver::shutdown ()
    at E:/git/fltk/master/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx:59
# 6  0x00007ff7d99d48bd in Fl_Win32_At_Exit::~Fl_Win32_At_Exit (this=0x7ff7d9a814b1 <win32_at_exit>,
    __in_chrg=<optimized out>) at E:/git/fltk/master/src/Fl_win32.cxx:626
# 7  0x00007ff7d993ba71 in __tcf_1 () at E:/git/fltk/master/src/Fl_win32.cxx:630
# 8  0x00007ffdf23da73b in msvcrt!_initterm_e () from C:\Windows\System32\msvcrt.dll
# 9  0x00007ff7d9889f25 in main (argc=3, argv=0x19022251810) at E:/git/fltk/master/fluid/fluid.cxx:1931
```

Stack frames `6` and `7` are obviously from code in FLTK, this is a step forward!

In case someone wants to take a look, this is from git commit 5a8b9d9130d9c9bd0adeca33ac051eb9249953fb.

Still investigating...

**Comment by erco77** [comment]
On 1/6/23 11:56, Albrecht Schlosser wrote:

> More results from gdb, looking around w/o a real plan, but ...
>
> |(gdb) bt # 0 0x00007ffdf382d144 in ntdll!ZwWaitForSingleObject () from C:\Windows\SYSTEM32\ntdll.dll # 1 0x00007ffdf0ea306e in 
> WaitForSingleObjectEx () from C:\Windows\System32\KernelBase.dll # 2 0x00007ffdcf5c8efa in gdiplus!GdiplusStartup () from 
> C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll # 3 0x00007ffdcf5c6bc4 in 
> gdiplus!GdiplusShutdown () from C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll 
> # 4 0x00007ffdcf5c6a43 in gdiplus!GdiplusShutdown () from 
> C:\Windows\WinSxS\amd64_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.19041.2251_none_91a40448cc8846c1\GdiPlus.dll # 5 0x00007ff7d992843b in 
> Fl_GDIplus_Graphics_Driver::shutdown () at E:/git/fltk/master/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx:59 # 6 0x00007ff7d99d48bd in 
> Fl_Win32_At_Exit::~Fl_Win32_At_Exit (this=0x7ff7d9a814b1 <win32_at_exit>, __in_chrg=<optimized out>) at E:/git/fltk/master/src/Fl_win32.cxx:626 # 7 
> 0x00007ff7d993ba71 in __tcf_1 () at E:/git/fltk/master/src/Fl_win32.cxx:630 # 8 0x00007ffdf23da73b in msvcrt!_initterm_e () from 
> C:\Windows\System32\msvcrt.dll # 9 0x00007ff7d9889f25 in main (argc=3, argv=0x19022251810) at E:/git/fltk/master/fluid/fluid.cxx:1931 |
>
> Stack frames |6| and |7| are obviously from code in FLTK, this is a step forward!
>
> In case someone wants to take a look, this is from git commit 5a8b9d9 <https://github.com/fltk/fltk/commit/5a8b9d9130d9c9bd0adeca33ac051eb9249953fb>.
>
> Still investigating...
>
     Huh, looks like it's hanging on exit when that singleton instance of Fl_Win32_At_Exit
     class self destructs, the destructor (orange) shuts down the fonts and stuff,
     and in this case, looks like it's stuck shutting down the GDI stuff (red).

     If the command line tool is also hanging, perhaps the startup isn't initializing
     something that the shutdown code needs to work correctly, and is hanging on
     objects it doesn't need to, or some such.

**Comment by Albrecht-S** [comment]
@erco77 Yes, that's correct. I tracked it down to `Fl_GDIplus_Graphics_Driver::shutdown()` which calls `Gdiplus::GdiplusShutdown()`. If I skip that call it finishes correctly.
Modified source:
```
void Fl_GDIplus_Graphics_Driver::shutdown() {
#if (1) // 1 = skip, 0 = call Gdiplus::GdiplusShutdown()
  printf("Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = %lu!\n", gdiplusToken);
#else
  printf("Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = %lu!\n", gdiplusToken);
  Gdiplus::GdiplusShutdown(Fl_GDIplus_Graphics_Driver::gdiplusToken);
#endif
}
```

Build log without GDI shutdown:
```
$ rm -rf test/*.cxx test/*.h bin/test/*.exe; time ninja -j12
[76/115] Generating fast_slow.cxx, fast_slow.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092046!
[79/115] Generating inactive.cxx, inactive.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092062!
[80/115] Generating keyboard_ui.cxx, keyboard_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092109!
[81/115] Generating mandelbrot_ui.cxx, mandelbrot_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092156!
[82/115] Generating resize.cxx, resize.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092453!
[83/115] Generating checkers_pieces.cxx, checkers_pieces.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092281!
[84/115] Generating preferences.cxx, preferences.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092250!
[85/115] Generating radio.cxx, radio.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092312!
[86/115] Generating tabs.cxx, tabs.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092312!
[87/115] Generating tree.cxx, tree.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122092531!
[98/115] Generating valuators.cxx, valuators.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122093859!
[100/115] Generating CubeViewUI.cxx, CubeViewUI.h
Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = 122094609!
[115/115] Linking CXX executable bin\test\CubeView.exe

real    0m31.988s
user    0m0.000s
sys     0m0.015s
```

Build with Gdiplus shutdown:
```
$ rm -rf test/*.cxx test/*.h bin/test/*.exe; time ninja -j12
[79/116] Generating tabs.cxx, tabs.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122234656!
[80/116] Generating tree.cxx, tree.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122234656!
[82/116] Generating valuators.cxx, valuators.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122234781!
[89/116] Linking CXX executable bin\test\tree.exe
```

**Build hangs here with 9 (nine) active fluid processes!**

Observations:
- 3 (of 12) processes claim to execute the shutdown
- 9 processes are hanging
- the first two tokens are the same value!

I'm not sure if any of these observations is helpful.

Another build with GDI shutdown, but `-j1` succeeds:
```
$ rm -rf test/*.cxx test/*.h bin/test/*.exe; time ninja -j1
[80/115] Generating checkers_pieces.cxx, checkers_pieces.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122572328!
[83/115] Generating fast_slow.cxx, fast_slow.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122573375!
[86/115] Generating inactive.cxx, inactive.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122574343!
[89/115] Generating keyboard_ui.cxx, keyboard_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122575468!
[92/115] Generating mandelbrot_ui.cxx, mandelbrot_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122576593!
[95/115] Generating preferences.cxx, preferences.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122577640!
[98/115] Generating radio.cxx, radio.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122578765!
[101/115] Generating resize.cxx, resize.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122579796!
[104/115] Generating tabs.cxx, tabs.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122580828!
[107/115] Generating tree.cxx, tree.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122581937!
[110/115] Generating valuators.cxx, valuators.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122583640!
[113/115] Generating CubeViewUI.cxx, CubeViewUI.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 122584812!
[115/115] Linking CXX executable bin\test\CubeView.exe

real    1m56.548s
user    0m0.000s
sys     0m0.015s
```

... whatever this means.

It's obvious that this "hang" wouldn't happen if we configured w/o GDIplus because this code wouldn't be executed.

I'm now looking into the details of the commit that "fixes" the issue. IIRC this commit removed an "early initialization" of widgets which would happen at static initialization time. This could cause a conflict here ("[static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof)"?).

Even if I don't know yet what **_exactly_** happens, I'm making progress...

**Comment by Albrecht-S** [comment]
@erco77 Greg, thanks for your "condensed" copy of the stack trace! Looking closer at this part (repeated here, read the comments bottom-up!):
```
# 1  in WaitForSingleObjectEx ()                  <- does this cause the deadlock?
# 2  in gdiplus!GdiplusStartup ()                 <- startup! WHAT DOES THIS HERE?
# 3  in gdiplus!GdiplusShutdown ()                <- shutdown (again!)
# 4  in gdiplus!GdiplusShutdown ()                <- shutdown
# 5  in Fl_GDIplus_Graphics_Driver::shutdown ()   <- our code
```

... seems to reveal a weird fact (reading bottom-up):
In `# 5` (`Fl_GDIplus_Graphics_Driver::shutdown()` we call `GdiplusShutdown` which in turn seems to call `GdiplusShutdown` (again, a recursion?) and then it's really weird because it seems to call `GdiplusStartup()` - yes, **_STARTUP()_** !

Finally this calls `WaitForSingleObjectEx()` which **_supposedly_** causes the deadlock.

OK, this really seems to indicate that something is (was in that older code) not initialized correctly.

**Comment by Albrecht-S** [comment]
Surprise, surprise!
Back at the "working" commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de and with the debug statements included, I see:
```
$ git log -1
commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de (HEAD)
Author: ManoloFLTK <41016272+ManoloFLTK@users.noreply.github.com>
Date:   Sat Oct 8 17:59:06 2022 +0200

    Wayland/X11 hybrid: use "bool fl_disable_wayland;" declaration.

$ ninja -j7
[530/574] Generating fast_slow.cxx, fast_slow.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[531/574] Generating inactive.cxx, inactive.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[532/574] Generating checkers_pieces.cxx, checkers_pieces.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[534/574] Generating keyboard_ui.cxx, keyboard_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[539/574] Generating mandelbrot_ui.cxx, mandelbrot_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[546/574] Generating preferences.cxx, preferences.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[550/574] Generating radio.cxx, radio.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[552/574] Generating resize.cxx, resize.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[555/574] Generating tabs.cxx, tabs.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[556/574] Generating tree.cxx, tree.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[557/574] Generating valuators.cxx, valuators.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[563/574] Generating CubeViewUI.cxx, CubeViewUI.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 0!
[574/574] Linking CXX executable bin\test\CubeView.exe
```

It works and (or because?) the "Token" is always 0 (zero)?

Next step follows...

**Comment by Albrecht-S** [comment]
OK, if anyone still follows: I added the "early graphics driver initialization" back in the constructor of `Fl_Widget` of the working commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de. I can confirm that this code caused the deadlock.

This was the minimal change that caused the issue, as if I was going back one (partial) commit. The result with test and debug statements is:
```
$ git log -1
commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de (HEAD)
Author: ManoloFLTK <41016272+ManoloFLTK@users.noreply.github.com>
Date:   Sat Oct 8 17:59:06 2022 +0200

    Wayland/X11 hybrid: use "bool fl_disable_wayland;" declaration.


$ git diff
diff --git a/src/Fl_Widget.cxx b/src/Fl_Widget.cxx
index 2289dcf0b..c4bb5d6f0 100644
--- a/src/Fl_Widget.cxx
+++ b/src/Fl_Widget.cxx
@@ -131,6 +131,14 @@ Fl_Widget::Fl_Widget(int X, int Y, int W, int H, const char* L) {

   parent_ = 0;
   if (Fl_Group::current()) Fl_Group::current()->add(this);
+
+#if (1) // EARLY INITIALIZATION - REMOVE THIS !   <<< (1) to simulate the previous commit
+  if (!fl_graphics_driver) {
+    // Make sure fl_graphics_driver is initialized. Important if we are called by a static initializer.
+    Fl_Display_Device::display_device();
+  }
+#endif // EARLY INITIALIZATION - REMOVE THIS !
+
 }

 void Fl_Widget::resize(int X, int Y, int W, int H) {
diff --git a/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx b/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx
index 188bd1d2d..3b6079c28 100644
--- a/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx
+++ b/src/drivers/GDI/Fl_GDI_Graphics_Driver.cxx
@@ -56,7 +56,12 @@ int Fl_GDIplus_Graphics_Driver::antialias() {
 ULONG_PTR Fl_GDIplus_Graphics_Driver::gdiplusToken = 0;

 void Fl_GDIplus_Graphics_Driver::shutdown() {
+#if (0) // 1 = skip, 0 = call Gdiplus::GdiplusShutdown()
+  printf("Fl_GDIplus_Graphics_Driver::shutdown() - skipping GdiplusShutdown, Token = %lu!\n", gdiplusToken);
+#else
+  printf("Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = %lu!\n", gdiplusToken);
   Gdiplus::GdiplusShutdown(Fl_GDIplus_Graphics_Driver::gdiplusToken);
+#endif
 }
 #endif

```

The result in the build process is as expected - it hangs in `fluid-cmd`:
```
$ rm -rf test/*.cxx test/*.h bin/test/*.exe; time ninja -j7
[80/117] Generating mandelbrot_ui.cxx, mandelbrot_ui.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125475687!
[84/117] Generating radio.cxx, radio.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125476531!
[87/117] Generating resize.cxx, resize.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125477109!
[90/117] Generating tabs.cxx, tabs.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125477812!
[93/117] Generating tree.cxx, tree.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125478312!
[96/117] Generating valuators.cxx, valuators.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125479171!
[99/117] Generating CubeViewUI.cxx, CubeViewUI.h
Fl_GDIplus_Graphics_Driver::shutdown() - executing GdiplusShutdown, Token = 125480343!
[102/117] Linking CXX executable bin\test\CubeView.exe
```
(hangs here). Q.E.D.

**Comment by Albrecht-S** [comment]
**_Summary:_**

- it is now clear that the "early graphics driver initialization" in the constructor of `Fl_Widget` had caused the "hang" if `GDIplus` was configured because the `GDIplus` shutdown caused a deadlock
- this "early graphics driver initialization" had been in FLTK for a very long time (longer than `GDIplus`)
- the "hang" appeared when both came together and ...
- it disappeared "magically" when the "early graphics driver initialization" was removed in git commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de
- it is not clear what exactly is missing or what is wrong with the initialization and shutdown **if** the deadlock happens

So, that's it for today for me. It's late here and it's been a long debugging session. I learned how to attach gdb to a running process on Windows and how to select threads in gdb to display the correct backtrace. This helped a lot to find the issue.

I also learned that I could "debug" a running process from within the Windows task manager (right click on a process, select "Debug") which would start the debugger in a Visual Studio instance - but I couldn't find the thread and backtrace info in the VS debugger. Hence I gave that up (keep using gdb).

I'm not sure if I will take another look tomorrow or later and if it's worth the effort, particularly since we found out that the issue does no longer exist.

Thanks to @J0hanss0n for finding the commit that "fixed" the issue and to all who helped as well.

**Comment by MatthiasWM** [comment]
So, um, I think the original code was added for Widget constructors that call innocent functions like fl_height() which in turn activates the display. This particular call was made for Widgets that are initialised before main() is called and, thanks to the "[static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof)" you already mentioned, are initialised before the display. 

I would check two things: when the code was removed, was the original problem fixed? And if the original problem was *not* fixed, what is the difference of calling this earlier than later, which in the end leads to the deadlock.

**Comment by Albrecht-S** [comment]
The "early initialization" of the graphics driver was introduced a long time ago. I found this out when I worked together with Manolo on the Wayland drivers. The problem was that nobody knew **why** this code had been included, except the comment saying that it was "**important** for static initialization of widgets" or something like that.

The problem we needed to solve was the hybrid FLTK library. At the time of static initialization the required driver was not yet defined, particularly because we needed to check environment variables and - at that time - the model was to optionally call a function in `main()` (i.e. after static initialization) - which is no longer the case (see `fl_disable_wayland`).

My point was that this driver initialization should not be there at all. Static initialization of widgets should **_never_** need to initialize or even call the graphics driver (in theory). After long discussions and fixing the Wayland initialization Manolo managed to make it work w/o this early initialization and the result was commit 4b945a3086011b6a59b5aef434cdbe8cda96c2de. We tried (tested) several constellations and could not find a real reason why this early initialization was necessary - if it was at some time, then it is supposedly no longer needed.

My gut feeling that it shouldn't be there seems to be confirmed by **this issue** and its "magic" resolution by removing the early driver initialization. Although we didn't **know** of this issue when we discussed the removal.

The question "_**what is the difference of calling this earlier than later, which in the end leads to the deadlock**_" is still not clear - WRT GDIplus initialization and shutdown. It's pretty clear that we don't need it in FLTK for static widget initialization.

**_[rant mode on]_**

I also wonder whether we really need to call `Gdiplus::GdiplusShutdown()`. This is one of the silly Windows problems: why do we need to call a shutdown() function for a drawing library we're using when the program exits anyway? That's one of the things I don't like - and understand - of Windows.

[From the documentation](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusinit/nf-gdiplusinit-gdiplusshutdown) (emphasis added by me):

"The GdiplusShutdown function cleans up resources used by Windows GDI+. Each call to GdiplusStartup **should be paired** with a call to GdiplusShutdown. [...] You must call GdiplusStartup before you create any GDI+ objects, and **you must delete all of your GDI+ objects (or have them go out of scope) before you call GdiplusShutdown**. [...] For Windows 7 and earlier, if GDI+ can't create a font family, it substitutes the generic Sans Serif family and client-side caches the pointer for the generic family. **Because calls to GdiplusStartup and GdiplusShutdown are not aware of this caching**, the operating system retains the pointer for the generic family object **past the object's lifetime**, which **causes the operating system to crash**."

End of documentation.

My comment: Really? The documentation says **"should"** and **"must"** but doesn't say what happens if you don't do it. The user is **required** to do this and that, but Windows 7 happily **crashes(!)** if a font can't be loaded. And nobody can complain because it's documented.

**_[rant mode off]_**

**Comment by erco77** [comment]
That sounds like an awesome denial of service hack.
Also, a fun way to quickly reboot the machine..!

**Comment by MatthiasWM** [comment]
OK, I am focussing now on making sure that for every GdiplusStartup, we call GdipulsShutdown. This is much trickier than I thought, and may even be impossible. I am basing my findings on https://learn.microsoft.com/en-gb/windows/win32/api/gdiplusinit/nf-gdiplusinit-gdiplusstartup?redirectedfrom=MSDN when linking dynamically.

So, neither startup nor shutdown can be called from DllMain. Well, static initialisation may already open the display, which means just that: DllMain may call GdiplusStartup. That's bad because GdiPlus.dll may not even be linked in yet. On Windows, dynamically linking is generally a bad idea anyway.

I will focus on statically linked apps and see if I can reproduce and solve the recursions observed by @Albrecht-S .

**Comment by Albrecht-S** [comment]
Thanks, @MatthiasWM, for continuing my investigation. After I had lost momentum I had one more thought which I'd like to share now. It's very vague but anyway...

In this special case of `fluid -c` it seems that we called (in these older commits with "early initialization") `GdiplusStartup` because we loaded the graphics driver (I assume it **did** call `GdiplusStartup`). However, then we didn't actually open the display and never created any Gdiplus **objects**. Finally we called `GDIplusShutdown` which magically called itself (some kind of recursion) and called `GdiplusStartup` (again), which produced the deadlock.

This didn't do any harm if it was all done sequentially, i.e. with `make -j1` or `ninja -j1` (no parallel building). But it seemed to cause the deadlock when run in a parallel build.

Here's **_my hypothesis_**: what if this is a bug in Windows (ehm, no, ... see [1])? We call `GdiplusStartup` and `GDIplusShutdown` w/o creating Gdiplus objects - and we do this in many processes in parallel! This is of course only speculation. It might cause the strange effect that two processes acquire the same GDIplus token (see my logs above) and this makes Windows choke for whatever reason. Finally it calls GdiplusStartup from within GDIplusShutdown and this causes the noticed deadlock.

My idea was to write a small program that does exactly this: call `GdiplusStartup`, do something, maybe wait a second, then call `GDIplusShutdown`. This program would have to be executed in parallel at least 3 to 5 times to see what happens. Does it create the same deadlock situation? This could even be used **instead** of our `fluid` as a simple demo program that copies or renames a file, if necessary (e.g. rename a required .cxx file from one dir to another dir to "simulate" `fluid -c`). That's only a thought though. Maybe this program could be launched by a shell script or another program or whatever. But since I lost steam I never wrote this program. Anyway, it's an idea you, Matthias, may pursue if you think it could be helpful.

Good luck!
---------------------------
[1] of course this is not a **Windows** bug, it's an invalid calling sequence of our program because we don't **use** GDIplus but call `GdiplusStartup` and `GDIplusShutdown` ;-) ;-)

**Comment by MatthiasWM** [comment]
Hopefully solved in #679

**Comment by Albrecht-S** [comment]
Unfortunately the given commits don't fix the underlying problem, the deadlock. This deadlock appears to exist only in MinGW builds - at least this is what I found out by running up to 200 parallel processes with `fluid-cmd` built by both MinGW and VS 2019. To verify that the problem still exists I patched the "early graphics driver initialization" code back into `Fl_Widget.cxx`. The only obvious difference in linked DLL's (as shown by `ldd`) are the runtime DLL's (all other DLL's are identical in both tested builds).

- MinGW:
```
       ucrtbased.dll => /c/Windows/System32/ucrtbased.dll (0x1aef0810000)
       ucrtbased.dll => /c/Windows/SYSTEM32/ucrtbased.dll (0x7ffa3b520000)
```

- VS 2019:
```
       VCRUNTIME140D.dll => /c/Windows/SYSTEM32/VCRUNTIME140D.dll (0x7ffa50960000)
       VCRUNTIME140_1D.dll => /c/Windows/SYSTEM32/VCRUNTIME140_1D.dll (0x7ffa4e2a0000)
```

It's noteworthy that I could not reproduce the issue with a simple test program that creates a window (thus initializing the graphics driver) and terminates w/o showing the window.
```C++
#include <FL/Fl_Window.H>
#include <Windows.h>
#include <stdio.h>

int msec = 500;

int main(int argc, char **argv) {
  Fl_Window *win = new Fl_Window(100, 100, 300, 300, "test");  // lädt Grafik-Treiber und ruft GdiplusStartup()
  // printf("Window created. Sleeping %d ms\n", msec);
  // fflush(stdout);
  Sleep(msec);
  return 0;
}
```

**Comment by Albrecht-S** [comment]
FTR: I set labels "fixed" and "wontfix" because the original issue has been fixed (doesn't exist in current git) but the underlying problem (deadlock in GdiplusShutdown()) could not be resolved and won't be worked on further. It might appear again in another context.
